### ISSUE TITLE
Subnet selection in Shared VPC architecture is not possible

### ISSUE THREAD

**Issue by irasnyd** [aws/karpenter#1368]
## Version

**Karpenter**: v0.6.3

**Kubernetes**: v1.21.5-eks-bc4871b

## Expected Behavior

When using a [Shared VPC architecture](https://aws.amazon.com/blogs/networking-and-content-delivery/vpc-sharing-a-new-approach-to-multiple-accounts-and-vpc-management/), the tags associated with each subnet are not available.

The Karpenter AWS Provider's `subnetSelector` field seems to require access to the subnet tags in order to discover subnets where instances can be launched.
 https://karpenter.sh/v0.6.3/aws/provisioning/#subnetselector

I expect to be able to specify a list of subnets by ID/ARN to handle the Shared VPC architecture.

## Actual Behavior

In a Shared VPC architecture, the tags on each subnet are not available. Therefore Karpenter cannot discover the subnets by tag. Therefore Karpenter cannot operate.

## Steps to Reproduce the Problem

* Create a multi-account AWS architecture, following the current AWS best practices
* Use one account to create a VPC and share with RAM
* Use another account to accept the Shared VPC
* Launch an EKS cluster in the Shared VPC account
* Attempt to use Karpenter, notice that it cannot discover subnets

The reason Karpenter cannot discover subnets by tag is because this is one of the [Limitations of VPC Sharing](https://docs.aws.amazon.com/vpc/latest/userguide/vpc-sharing.html#:~:text=VPC%20tags%2C%20and%20tags%20for%20the%20resources%20within%20the%20shared%20VPC%20are%20not%20shared%20with%20the%20participants.)

## Resource Specs and Logs

Here is the output of `aws ec2 describe-subnets` inside the account that has accepted the Shared VPC:

```
{
    "Subnets": [
        {
            "AvailabilityZone": "us-west-2b",
            "AvailabilityZoneId": "usw2-az1",
            "AvailableIpAddressCount": 65452,
            "CidrBlock": "10.152.0.0/16",
            "DefaultForAz": false,
            "MapPublicIpOnLaunch": true,
            "MapCustomerOwnedIpOnLaunch": false,
            "State": "available",
            "SubnetId": "subnet-0fa63929b62040515",
            "VpcId": "vpc-0eb4edcf4e1278015",
            "OwnerId": "012345678901",
            "AssignIpv6AddressOnCreation": false,
            "Ipv6CidrBlockAssociationSet": [],
            "SubnetArn": "arn:aws:ec2:us-west-2:012345678901:subnet/subnet-0fa63929b62040515",
            "EnableDns64": false
        },
        {
            "AvailabilityZone": "us-west-2a",
            "AvailabilityZoneId": "usw2-az2",
            "AvailableIpAddressCount": 65509,
            "CidrBlock": "10.0.0.0/16",
            "DefaultForAz": false,
            "MapPublicIpOnLaunch": false,
            "MapCustomerOwnedIpOnLaunch": false,
            "State": "available",
            "SubnetId": "subnet-019ebd8101c5c62f3",
            "VpcId": "vpc-0eb4edcf4e1278015",
            "OwnerId": "012345678901",
            "AssignIpv6AddressOnCreation": false,
            "Ipv6CidrBlockAssociationSet": [],
            "SubnetArn": "arn:aws:ec2:us-west-2:012345678901:subnet/subnet-019ebd8101c5c62f3",
            "EnableDns64": false
        },
        {
            "AvailabilityZone": "us-west-2d",
            "AvailabilityZoneId": "usw2-az4",
            "AvailableIpAddressCount": 65482,
            "CidrBlock": "10.154.0.0/16",
            "DefaultForAz": false,
            "MapPublicIpOnLaunch": true,
            "MapCustomerOwnedIpOnLaunch": false,
            "State": "available",
            "SubnetId": "subnet-08fa1c0a8d2e3d950",
            "VpcId": "vpc-0eb4edcf4e1278015",
            "OwnerId": "012345678901",
            "AssignIpv6AddressOnCreation": false,
            "Ipv6CidrBlockAssociationSet": [],
            "SubnetArn": "arn:aws:ec2:us-west-2:012345678901:subnet/subnet-08fa1c0a8d2e3d950",
            "EnableDns64": false
        },
        {
            "AvailabilityZone": "us-west-2b",
            "AvailabilityZoneId": "usw2-az1",
            "AvailableIpAddressCount": 65509,
            "CidrBlock": "10.1.0.0/16",
            "DefaultForAz": false,
            "MapPublicIpOnLaunch": false,
            "MapCustomerOwnedIpOnLaunch": false,
            "State": "available",
            "SubnetId": "subnet-0471ba2f8b8a118bf",
            "VpcId": "vpc-0eb4edcf4e1278015",
            "OwnerId": "012345678901",
            "AssignIpv6AddressOnCreation": false,
            "Ipv6CidrBlockAssociationSet": [],
            "SubnetArn": "arn:aws:ec2:us-west-2:012345678901:subnet/subnet-0471ba2f8b8a118bf",
            "EnableDns64": false
        },
        {
            "AvailabilityZone": "us-west-2c",
            "AvailabilityZoneId": "usw2-az3",
            "AvailableIpAddressCount": 65510,
            "CidrBlock": "10.2.0.0/16",
            "DefaultForAz": false,
            "MapPublicIpOnLaunch": false,
            "MapCustomerOwnedIpOnLaunch": false,
            "State": "available",
            "SubnetId": "subnet-044343e27b7ad9fa2",
            "VpcId": "vpc-0eb4edcf4e1278015",
            "OwnerId": "012345678901",
            "AssignIpv6AddressOnCreation": false,
            "Ipv6CidrBlockAssociationSet": [],
            "SubnetArn": "arn:aws:ec2:us-west-2:012345678901:subnet/subnet-044343e27b7ad9fa2",
            "EnableDns64": false
        },
        {
            "AvailabilityZone": "us-west-2a",
            "AvailabilityZoneId": "usw2-az2",
            "AvailableIpAddressCount": 65495,
            "CidrBlock": "10.151.0.0/16",
            "DefaultForAz": false,
            "MapPublicIpOnLaunch": true,
            "MapCustomerOwnedIpOnLaunch": false,
            "State": "available",
            "SubnetId": "subnet-02f8d1efd38e4851a",
            "VpcId": "vpc-0eb4edcf4e1278015",
            "OwnerId": "012345678901",
            "AssignIpv6AddressOnCreation": false,
            "Ipv6CidrBlockAssociationSet": [],
            "SubnetArn": "arn:aws:ec2:us-west-2:012345678901:subnet/subnet-02f8d1efd38e4851a",
            "EnableDns64": false
        },
        {
            "AvailabilityZone": "us-west-2c",
            "AvailabilityZoneId": "usw2-az3",
            "AvailableIpAddressCount": 65464,
            "CidrBlock": "10.153.0.0/16",
            "DefaultForAz": false,
            "MapPublicIpOnLaunch": true,
            "MapCustomerOwnedIpOnLaunch": false,
            "State": "available",
            "SubnetId": "subnet-01e5a78553636f300",
            "VpcId": "vpc-0eb4edcf4e1278015",
            "OwnerId": "012345678901",
            "AssignIpv6AddressOnCreation": false,
            "Ipv6CidrBlockAssociationSet": [],
            "SubnetArn": "arn:aws:ec2:us-west-2:012345678901:subnet/subnet-01e5a78553636f300",
            "EnableDns64": false
        },
        {
            "AvailabilityZone": "us-west-2d",
            "AvailabilityZoneId": "usw2-az4",
            "AvailableIpAddressCount": 65531,
            "CidrBlock": "10.3.0.0/16",
            "DefaultForAz": false,
            "MapPublicIpOnLaunch": false,
            "MapCustomerOwnedIpOnLaunch": false,
            "State": "available",
            "SubnetId": "subnet-0dd5ed75583afcdb5",
            "VpcId": "vpc-0eb4edcf4e1278015",
            "OwnerId": "012345678901",
            "AssignIpv6AddressOnCreation": false,
            "Ipv6CidrBlockAssociationSet": [],
            "SubnetArn": "arn:aws:ec2:us-west-2:012345678901:subnet/subnet-0dd5ed75583afcdb5",
            "EnableDns64": false
        }
    ]
}
```

**Comment by irasnyd** [comment]
I am working on this issue here: https://github.com/aws/karpenter/compare/main...irasnyd:subnet-arn?expand=1

I'll make a PR when finished. In the mean time, comments on the approach are appreciated.

**Comment by ellistarn** [comment]
You can use names or tags for discovery. Can you elaborate on how this is insufficient for the shared VPC use case. 
```
subnetSelector:
  Name: *wildcard*
```

**Comment by irasnyd** [comment]
Hi @ellistarn,

Thanks for looking into this with me. I tried the solution you proposed, however I end up getting this message logged over and over: https://github.com/aws/karpenter/blob/main/pkg/cloudprovider/aws/subnets.go#L57

I read through the code, and figured out that the solution you proposed ends up creating this filter: https://github.com/aws/karpenter/blob/main/pkg/cloudprovider/aws/subnets.go#L74-L77

However, one of the limitations of a [Shared VPC Architecture](https://docs.aws.amazon.com/vpc/latest/userguide/vpc-sharing.html#:~:text=VPC%20tags%2C%20and%20tags%20for%20the%20resources%20within%20the%20shared%20VPC%20are%20not%20shared%20with%20the%20participants.) is:
> VPC tags, and tags for the resources within the shared VPC are not shared with the participants.

This means that I cannot use _any tags at all_ (including the `Name` tag) to locate the subnets. The participant vpc **cannot read any tags** from the subnets. The participant vpc has access to the the subnet id and arn, but not the tags. I attached the output of `aws ec2 describe-subnets` down below that demonstrates this fact.

I tested using this YAML:

```yaml
subnetSelector:
  Name: "*public*"
```

It doesn't match any subnets because it cannot read the tags at all. Take a careful look at the following output from `aws ec2 describe-subnets`. It doesn't contain any tags at all! They aren't available in a Shared VPC architecture.

```json
{
    "Subnets": [
        {
            "AvailabilityZone": "us-west-2b",
            "AvailabilityZoneId": "usw2-az1",
            "AvailableIpAddressCount": 65452,
            "CidrBlock": "10.152.0.0/16",
            "DefaultForAz": false,
            "MapPublicIpOnLaunch": true,
            "MapCustomerOwnedIpOnLaunch": false,
            "State": "available",
            "SubnetId": "subnet-0fa63929b62040515",
            "VpcId": "vpc-0eb4edcf4e1278015",
            "OwnerId": "012345678901",
            "AssignIpv6AddressOnCreation": false,
            "Ipv6CidrBlockAssociationSet": [],
            "SubnetArn": "arn:aws:ec2:us-west-2:012345678901:subnet/subnet-0fa63929b62040515",
            "EnableDns64": false
        }
    ]
}
```

**Comment by ellistarn** [comment]
Interesting -- this is news to me. I could imagine building in a magic selector the logic to convert Arn or ID into the corresponding requests.

**Comment by ellistarn** [comment]
I'd make it a magic field of the map. Arn is a slightly more robust concept than SubnetId, I think. e.g.

```
subnetSelector:
  Arn: "arn:aws:ec2:us-west-2:012345678901:subnet/subnet-0fa63929b62040515"
securityGroupSelector:
  Arn: "arn:aws:ec2:us-west-2:012345678901:securityGroup/foobar"
```

I can't help but worry that this makes multi-az very hard. How are you going to achieve this? Multiple provisioners?

**Comment by irasnyd** [comment]
Here is an implementation of the idea: https://github.com/aws/karpenter/compare/main...irasnyd:subnet-selector-magic-tags?expand=1

**Comment by ellistarn** [comment]
I'm very hesitant towards this, since it's magic/complexity/api surface that must be maintained. I think I need to understand the shared VPC feature better and why it has the restrictions it does.

**Comment by irasnyd** [comment]
No problem; I definitely understand the concern about complexity.

How can I help share my knowledge about the Shared VPC feature? As I understand the design, this was a way that AWS designed to be able to share network infrastructure resources (NAT Gateways, IGW Internet Gateways, VPN Connections, Direct Connects, etc.) across many accounts in the AWS Multi-Account Architecture. This provides an alternative to using Transit Gateway to link many VPCs (Transit Gateway is somewhat expensive: $0.02/GB of traffic.) A Shared VPC Architecture also provides a way for companies which have a dedicated "Network Team" to manage the network resources for the company as a whole.

As I understand it, the vpc and subnet tags are not shared to the participant accounts at all, by design. I think this decision was made as a security feature: to keep tags private within the account where the resources they are attached to are created. In case there is "private" information in the tags that shouldn't leak out the participant accounts.

* [AWS Shared VPC Limitations Documentation](https://docs.aws.amazon.com/vpc/latest/userguide/vpc-sharing.html#:~:text=VPC%20tags%2C%20and%20tags%20for%20the%20resources%20within%20the%20shared%20VPC%20are%20not%20shared%20with%20the%20participants.)
* [Using VPC Sharing for a Cost-Effective Multi-Account Microservice Architecture](https://aws.amazon.com/blogs/architecture/using-vpc-sharing-for-a-cost-effective-multi-account-microservice-architecture/)
* [VPC sharing: A new approach to multiple accounts and VPC management
](https://aws.amazon.com/blogs/networking-and-content-delivery/vpc-sharing-a-new-approach-to-multiple-accounts-and-vpc-management/)

**Comment by ellistarn** [comment]
> I think this decision was made as a security feature: to keep tags private within the account where the resources they are attached to are created. In case there is "private" information in the tags that shouldn't leak out the participant accounts.

I might be worth reaching out and asking for a workaround from the VPC side. One option could be a separate view of tags from each account. Or potentially allowing tags under a certain prefix come through the API.

**Comment by vtomasr5** [comment]
Hi
This should be similar to the [AWS ALB controller](https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.4/guide/ingress/annotations/#subnets) subnets where you have to put the list for the subnetIds.
We're using Shared VPC and this works for the ALB controller because tagging the network coming from another account didn't work for us for the ALB controller.

**Comment by irasnyd** [comment]
Fixed by #1732
